File: packages/python/client/starfish_sdk/entitlements.py
"""Client-side helper for reading a user's entitlement document."""

from __future__ import annotations

from starfish_sdk.client import StarfishClient
from starfish_sdk.types import StarfishHttpError


async def pull_entitlements(
    client: StarfishClient,
    user_id: str,
    path: str = "/pull/users/{user_id}/entitlements",
    field: str = "features",
) -> list[str]:
    """Fetch the list of feature slugs from a user's entitlement document.

    Returns an empty list if the document does not exist yet or the field is
    absent — callers never need to handle a 404.

    ::

        from starfish_sdk import pull_entitlements

        features = await pull_entitlements(client, user_id)
        # e.g. ["premium-package-1", "paid-cloud-sync"]

        if "paid-cloud-sync" in features:
            # unlock cloud sync UI
            pass

    :param client: An authenticated :class:`~starfish_sdk.client.StarfishClient`.
    :param user_id: The identity segment used in the entitlement collection path.
    :param path: Path template; ``{user_id}`` is replaced with ``user_id``.
        Defaults to ``"/pull/users/{user_id}/entitlements"``.
    :param field: Field name inside the document ``data`` object that holds the
        feature slug list.  Defaults to ``"features"``.
    :returns: List of feature slug strings, or an empty list if none found.
    """
    resolved_path = path.replace("{user_id}", user_id)
    try:
        result = await client.pull(resolved_path)
        data = result.data if isinstance(result.data, dict) else {}
        feature_list = data.get(field)
        if not isinstance(feature_list, list):
            return []
        return [f for f in feature_list if isinstance(f, str)]
    except StarfishHttpError as exc:
        if exc.status == 404:
            return []
        raise

File: packages/ts/entitlements/src/entitlements.ts
import type { StarfishClient } from "@drakkar.software/starfish-client"
import { StarfishHttpError } from "@drakkar.software/starfish-client"

export interface PullEntitlementsOptions {
  /**
   * Path template for the entitlement document.
   * `{userId}` is replaced with the `userId` argument.
   * Defaults to `"/pull/users/{userId}/entitlements"`.
   */
  path?: string
  /**
   * Field name in the document `data` object that holds the feature slug array.
   * Defaults to `"features"`.
   */
  field?: string
}

/**
 * Fetches the list of feature slugs from a user's entitlement document.
 *
 * Returns an empty array if the document does not exist yet or the features
 * field is absent — so callers never need to handle a 404.
 *
 * ```ts
 * import { pullEntitlements } from "@drakkar.software/starfish-client"
 *
 * const features = await pullEntitlements(client, userId)
 * // e.g. ["premium-package-1", "paid-cloud-sync"]
 *
 * if (features.includes("paid-cloud-sync")) {
 *   // unlock cloud sync UI
 * }
 * ```
 *
 * The path template must match the server-side collection's `storagePath`.
 * With the recommended default config:
 * ```ts
 * { storagePath: "users/{identity}/entitlements" }
 * // → path: "/pull/users/{userId}/entitlements"  (default)
 * ```
 */
export async function pullEntitlements(
  client: StarfishClient,
  userId: string,
  opts?: PullEntitlementsOptions,
): Promise<string[]> {
  const path = (opts?.path ?? "/pull/users/{userId}/entitlements").replace("{userId}", userId)
  const field = opts?.field ?? "features"

  try {
    const result = await client.pull(path)
    const list = (result.data as Record<string, unknown> | null)?.[field]
    if (!Array.isArray(list)) return []
    return list.filter((s): s is string => typeof s === "string")
  } catch (err) {
    if (err instanceof StarfishHttpError && err.status === 404) return []
    throw err
  }
}
